### DISCUSSION TITLE
Question regarding related records with non primary record ids.

### DISCUSSION THREAD

**Discussion by Trini8ed** [pocketbase/pocketbase#1895] [Q&A]
Should be a fairly quick question, I'm relating documents that don't use the standard primary key and want to make sure I'm not missing anything to get them to link up correctly. Seems the front-end UI is not linking them correctly, or perhaps I did something wrong when writing the records in-case I missed something special (Record made from a custom model).

You can see below its missing the relation after clicking on the record, but it does show the correct ID that I'm referencing to. Any ideas here?

``` go
type Config struct {
	models.BaseModel

	ConfigID     string `db:"config_id" json:"id"`
	MaterialID   string `db:"material_id" json:"materialId"`
}

configRecord := &Config{
	ConfigID:     config.ID, // UUID string
	MaterialID:   material.ID, // UUID string
}
if err := app.Dao().Save(configRecord); err != nil {
	logger.ErrorLog.Println(err.Error())
}
```

![image](https://user-images.githubusercontent.com/1361448/220591240-99ee02f1-1a8a-4052-aefd-6020ffb953e7.png)

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand the screenshot.

What is the relation between the shown `configs` and `finishes` collections? It seems that they have the same fields or `finishes` was renamed to `configs`?

The "Display fields" is only for display purposes in the Admin UI. It is not used for relating records. Records are related only via their `id` field.

As long as there is a `materials` record with the provided id (even if it is a custom one), the relation should work just fine.

If that doesn't help, could  you provide a more complete code sample and details about your collections structure?

**Reply by ganigeorgiev** [reply]
Also note that the struct tags of the `Config` model struct are "incorrect". You shouldn't "replace" the original return id in the json with the custom one (_unless you really want this behavior_). If you want to use custom ids as primary keys, you can directly assign them to your `id` field when creating the model.

**Reply by Trini8ed** [reply]
Let me give a more simplified example that is reproducible on your end. There are more relations going on other than just configs and finishes. I'm going to abstract away to just materials and configs for the example below.

In this case, config_id and material_id are not unique and there can be multiple records with the same config and material ids. This is why I still need to leave the pocket base generated id alone. I'm doing a bit of weird implementation to query records based on date and the config_id or material_id.

When you create a relation like this for some reason the Web UI doesn't show them as linked. When you click on any of the records created pragmatically they show as blank above the "Open Picker" (similar to what's shown in the Orange box in the screenshot, Red box is what I would expect). The main table view of all the records shows the correct material_id (They are not blank) to what I set programmatically, but it seems to not be linking to the record like If I did this manually through the Web UI. I can't tell if this is just a Web UI bug or if they are not actually linked under the hood. If you create the relations manually through the Web UI it worked as expected.

If you're linking to multiple different records of the same material_id in this case is this the expected behavior?

Steps to reproduce:
1) Create 2 collections 'materials' and 'configs'
2) 'materials' collection needs 1 TEXT field called material_id
3) 'configs' needs 1 TEXT field called config_id and 1 RELATION field called material_id that references a material_id from the 'materials' collection.
4) Using pocket base as a framework add the following record creation to the OnBeforeServe hook method to generate the docs pragmatically.
5) You can see that when clicking on the newly created record in the 'configs' collection through the Web UI it shows as blank.

```go

var _ models.Model = (*Material)(nil)
var _ models.Model = (*Config)(nil)

func (m *Material) TableName() string {
	return "materials"
}

func (m *Config) TableName() string {
	return "configs"
}

type Material struct {
	models.BaseModel

	MaterialID string `db:"material_id"`
}

type Config struct {
	models.BaseModel

	ConfigID     string `db:"config_id"`
	MaterialID   string `db:"material_id"`
}

// ...

materialRecord1 := &Material{
	MaterialID:   "e9c3f47e-b2ec-11ed-afa1-0242ac120002", // UUID string
}

materialRecord2 := &Material{
	MaterialID:   "e9c3f47e-b2ec-11ed-afa1-0242ac120002", // UUID string
}

if err := app.Dao().Save(materialRecord1); err != nil {
	log.Fatal(err)
}
if err := app.Dao().Save(materialRecord2); err != nil {
	log.Fatal(err)
}

configRecord := &Config{
	ConfigID:     "f22a4df2-b2ec-11ed-afa1-0242ac120002", // UUID string
	MaterialID:   "edf6c3be-b2ec-11ed-afa1-0242ac120002", // UUID string
}

if err := app.Dao().Save(configRecord); err != nil {
	log.Fatal(err)
}
```

**Reply by ganigeorgiev** [reply]
> 3. 'configs' needs 1 TEXT field called config_id and **1 RELATION field called material_id that references a material_id from the 'materials' collection**.

You can't do the bolded part. The reason for that is because the relations are only referenced via their PocketBase record id. In other words, the relation must be assigned to `materials.id`.

> If you create the relations manually through the Web UI it worked as expected.

This is because the Admin UI associates the relations via the PocketBase record id. You are seeing the `materials.material_id` value because that is the "Display fields" that you've selected ("Display fields" is only for display purposes in the Admin UI; the relation internally is still referenced by the PocketBase record id).

---

The above snippet should work if you assign the `Config.MaterialID` to `materialRecord1.Id` or `materialRecord2.Id`.

**Reply by Trini8ed** [reply]
I see so references are limited to the record.id itself, I'll just need to create extra fields like material_ref on top of materal_id if I want to link to also link it by reference.

Something like...
```go

var _ models.Model = (*Material)(nil)
var _ models.Model = (*Config)(nil)

func (m *Material) TableName() string {
	return "materials"
}

func (m *Config) TableName() string {
	return "configs"
}

type Material struct {
	models.BaseModel

	MaterialID string `db:"material_id"`
}

type Config struct {
	models.BaseModel

	ConfigID     string `db:"config_id"`
	MaterialID   string `db:"material_id"`
	MaterialRef string `db:"material_ref"`
}

// ...

materialRecord1 := &Material{
	MaterialID:   "e9c3f47e-b2ec-11ed-afa1-0242ac120002", // UUID string
}

materialRecord2 := &Material{
	MaterialID:   "e9c3f47e-b2ec-11ed-afa1-0242ac120002", // UUID string
}

if err := app.Dao().Save(materialRecord1); err != nil {
	log.Fatal(err)
}
if err := app.Dao().Save(materialRecord2); err != nil {
	log.Fatal(err)
}

configRecord := &Config{
	ConfigID:     "f22a4df2-b2ec-11ed-afa1-0242ac120002", // UUID string
	MaterialID:   "edf6c3be-b2ec-11ed-afa1-0242ac120002", // UUID string
	MaterialRef:   materialRecord2.Id,
}

if err := app.Dao().Save(configRecord); err != nil {
	log.Fatal(err)
}
```

These records will all be created at different times which is how I'll be referencing them to distinguish between materialRecord1 and materialRecord2 in case it's a bit confusing why I'm doing it this way. I have another data structure that refers to these docs by their material_id and it's not something I can change I need to look up records based on that id.

Are there any plans to allow for references by means other than the primary record.id?

**Reply by ganigeorgiev** [reply]
> Are there any plans to allow for references by means other than the primary record.id?

No, at least for now I don't have plans for this since it will complicate too much the internals.
Also keep in mind that, as far as I understand your structure, you have multiple records (`materials`) sharing the same ref key and even if we allow referencing by field different than the id, I'm not sure how this will even work since it will not be clear which record the field refer to and how the rels expand and filters will work.

I don't have a full picture of the use case and I'm not sure what to recommend, but one alternative approach could be to create an extra collection, eg. `materialRefs` that will have the custom id(s) and point the other collections to it.
